### PR TITLE
SONARIAC-1451 Properties grammar should accept key that contains comments indicators

### PR DESCRIPTION
**Pull request by mstachniuk**

### CODE REVIEW

File: iac-extensions/spring-config/src/main/antlr/PropertiesLexer.g4 [hunk 1]
```diff
@@ -7,7 +7,7 @@ public PropertiesLexer(CharStream input, boolean crLexerCostructor) {
 }
 }
 
-COMMENT   : [!#] -> pushMode(VALUE_MODE);
+COMMENT   : {this.getCharPositionInLine() == 0}? [!#] -> pushMode(VALUE_MODE);
```
Comment (petertrr) [inline iac-extensions/spring-config/src/main/antlr/PropertiesLexer.g4:10]: Will it cover lines starting with spaces and then comment?
```properties
foo=bar
    # comment
```

File: iac-extensions/spring-config/src/main/antlr/PropertiesParser.g4 [hunk 2]
```diff
@@ -9,8 +9,8 @@ propertiesFile
     ;
 
 row
-    : comment
-    | line
+    : line
+    | comment
```
Comment (petertrr) [inline iac-extensions/spring-config/src/main/antlr/PropertiesParser.g4:13]: How does this reordering relate to the new condition in lexer? Would it still work without reordering, if the lexer now matches comments only at the beginning of a line?

File: iac-extensions/spring-config/src/test/java/org/sonar/iac/springconfig/parser/properties/PropertiesParserBaseVisitorTest.java [hunk 3]
```diff
@@ -590,6 +592,38 @@ void shouldParseKeyEmptyValue() {
       "visitEol <EOF>");
   }
 
+  @ParameterizedTest
+  @CsvSource({"foo#,foo#",
+    "foo#=,foo#",
+    "foo!,foo!",
+    "foo!=,foo!"
```
Comment (petertrr) [inline iac-extensions/spring-config/src/test/java/org/sonar/iac/springconfig/parser/properties/PropertiesParserBaseVisitorTest.java:599]: Maybe also add cases when value is present?

File: iac-extensions/spring-config/src/test/java/org/sonar/iac/springconfig/parser/properties/PropertiesParserBaseVisitorTest.java [hunk 4]
```diff
@@ -590,6 +592,38 @@ void shouldParseKeyEmptyValue() {
...
+    assertThat(visitor.visited()).containsExactly(
+      "visitPropertiesFile %s<EOF><EOF>".formatted(code),
+      "visitRow %s<EOF>".formatted(code),
+      "visitLine %s<EOF>".formatted(code),
+      "visitKey %s".formatted(key),
+      "visitEol <EOF>");
+  }
+
+  @Test
+  void shouldParseKeyThatContainsHashAndEmptyValue() {
```
Comment (petertrr) [inline iac-extensions/spring-config/src/test/java/org/sonar/iac/springconfig/parser/properties/PropertiesParserBaseVisitorTest.java:613]: Is the name correct? The value is not empty here

**Review by petertrr — APPROVED** [review]
LGTM, please see my questions above

File: iac-extensions/spring-config/src/test/java/org/sonar/iac/springconfig/parser/properties/PropertiesParserBaseVisitorTest.java (hunk 4 above)
Comment (mstachniuk) [reply iac-extensions/spring-config/src/test/java/org/sonar/iac/springconfig/parser/properties/PropertiesParserBaseVisitorTest.java:613]: Sure, I will rename it.

File: iac-extensions/spring-config/src/test/java/org/sonar/iac/springconfig/parser/properties/PropertiesTestUtils.java [hunk 5]
```diff
@@ -29,6 +31,13 @@ public static PropertiesParser.PropertiesFileContext createPropertiesFileContext
     var propertiesLexer = new PropertiesLexer(inputCode, true);
     var commonTokenStream = new CommonTokenStream(propertiesLexer);
     var parser = new PropertiesParser(commonTokenStream);
+
+    // printing tokens for debugging
+    commonTokenStream.fill();
+    for (Token token : commonTokenStream.getTokens()) {
+      System.out.println(((CommonToken) token).toString(parser));
```
Comment (mstachniuk) [inline iac-extensions/spring-config/src/test/java/org/sonar/iac/springconfig/parser/properties/PropertiesTestUtils.java:38]: As this is a test code IMO it is ok to print to stdout

File: iac-extensions/spring-config/src/main/antlr/PropertiesLexer.g4 [hunk 6]
```diff
@@ -7,13 +7,19 @@ public PropertiesLexer(CharStream input, boolean crLexerCostructor) {
...
 LEADING_SPACING : {this.getCharPositionInLine() == 0}? [ \t\f\r\n\u2028\u2029]+ -> channel(HIDDEN);
 NEWLINE         : [\r\n\u2028\u2029]+;
 DELIMITER       : [ ]* [:=\t\f ] [ ]* -> pushMode(VALUE_MODE);
 SLASH           : '\\' -> more, pushMode(INSIDE);
 CHARACTER       : ~ [:=\r\n\u2028\u2029];
 
+mode COMMENT_MODE;
+
+COMMENT_NEW_LINE  : [\r\n\u2028\u2029]+      -> type(NEWLINE), popMode;
+COMMENT_DELIMITER : [ ]* [:=\t\f ] [ ]*      -> type(DELIMITER), popMode;
```
Comment (petertrr) [inline iac-extensions/spring-config/src/main/antlr/PropertiesLexer.g4:20]: Why do we need to `popMode` after delimiters in comments? Shouldn't it happen only after new lines?

**Pull request merged**